bin/sl-meshctl.js

```diff
@@ -123,213 +125,332 @@ function unknown() {
+          dieIf(err);
+          console.log('CPU profile written to `%s`, load into Chrome Dev Tools',
+            fileName);
+        });
+      });
+    }
+  );
 }
 
-function cmdHeapSnapshot(instance) {
+function cmdHeapSnapshot(client) {
   var target = mandatory('target');
   var prefix = optional(util.format('node.%s', target));
   var fileName = prefix + '.heapsnapshot';
 
-  instance.heapSnapshot(target, function(err, response) {
-    dieIf(err);
-    var profileId = response.profileId;
-    download(instance, profileId, fileName, function(err) {
+  client.resolveTarget(target,
+    function(err, service, executor, instance, process) {
       dieIf(err);
-      console.log('Heap snapshot written to `%s`, load into Chrome Dev Tools',
-        fileName);
-    });
-  });
+      process.heapSnapshot(function(err, response) {
+        dieIf(err);
+        var profileId = response.profileId;
+        download(instance, profileId, fileName, function(err) {
+          dieIf(err);
+          console.log(
+            'Heap snapshot written to `%s`, load into Chrome Dev Tools',
+            fileName
+          );
+        });
+      });
+    }
+  );
 }
 
-function cmdLs(instance) {
+function cmdLs(client) {
+  var targetService = mandatory('service');
   var depth = optional(Number.MAX_VALUE);
 
-  instance.npmModuleList(function(err, response) {
-    dieIf(err);
-    console.log(npmls.printable(response, depth));
-  });
+  client.serviceFind(targetService,
+    function(err, service) {
+      dieIf(err);
+      service.instances({limit: 1}, function(err, instances) {
+        dieIf(err);
+        if (instances.length !== 1) dieIf(Error('Invalid service'));
+        var instance = instances[0];
+        instance.npmModuleList(function(err, response) {
+          dieIf(err);
+          console.log(npmls.printable(response, depth));
+        });
+      });
+    });
 }
 
-function cmdPatch(instance) {
+function cmdPatch(client) {
   var target = mandatory('target');
   var patchFile = mandatory('patch file');
 
   var patchData = JSON.parse(fs.readFileSync(patchFile));
-  instance.applyPatch(target, patchData, function(err /*, response*/) {
-    dieIf(err);
-  });
+  client.resolveTarget(target,
+    function(err, service, executor, instance, process) {
+      dieIf(err);
+      process.applyPatch(patchData, function(err /*, response*/) {
+        dieIf(err);
+      });
+    }
+  );
 }
 
-function cmdEnvSet(instance) {
+function cmdEnvSet(client) {
+  var targetService = mandatory('service');
   var vars = mandatorySome('K=V');
   var env = _.reduce(vars, extractKeyValue, {});
-  instance.envSet(env, function(err, response) {
+
+  client.serviceFind(targetService, function(err, service) {
     dieIf(err);
-    console.log('Environment updated: %j', response.message);
+    service.setEnvs(env, function(err, responses) {
+      dieIf(err);
+      // Fail if any of the instances failed to update
```

Comment:
Hm, yes. I would prefer to leave this in here and add an XXX for when we add rollback support on server. The current way its implemented just delegates the update to each instance.